File: PaMVPDemo/lib-core/src/main/java/com/pa/core/base/BaseActivity.java
package com.pa.core.base;

import android.content.DialogInterface;
import android.content.DialogInterface.OnCancelListener;
import android.os.Bundle;
import android.support.annotation.Nullable;
import android.support.v7.app.AppCompatActivity;


import com.pa.core.view.CustomProgressDialog;

import butterknife.ButterKnife;
import butterknife.Unbinder;

/**
 * Created by wen on 2018/5/14.
 */
public abstract class BaseActivity<T extends IPresenter> extends AppCompatActivity implements IView {

    protected T mPresenter;
    private Unbinder unbinder;

    @Override
    protected void onCreate(@Nullable Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(getlayoutId());
        mPresenter = (T) createPresenter();
        unbinder = ButterKnife.bind(this);
        initData();
        initView();
        initTitle();
    }

    protected abstract int getlayoutId();

    protected abstract IPresenter createPresenter();

    public void initData() {
    }

    public void initView() {
    }

    public void initTitle(){
    }


    @Override
    public void showLoadingView() {
        if (!this.isFinishing()&&!CustomProgressDialog.isShow()) {
            CustomProgressDialog.show(this, "登录中...", false, onCancelListener);
        }
    }

    @Override
    public void hideLoadingView() {
        if (!this.isFinishing()&&CustomProgressDialog.isShow()) {
            CustomProgressDialog.dimiss();
        }
    }

    @Override
    protected void onDestroy() {
        if (null != mPresenter) {
            mPresenter.onDestroy();
        }
        if (null != unbinder) {
            unbinder.unbind();
            unbinder = null;
        }
        super.onDestroy();
    }

    /**
     * 监听dialog的返回  取消后注销请求
     */
    OnCancelListener onCancelListener = new OnCancelListener() {
        @Override
        public void onCancel(DialogInterface dialog) {
            mPresenter.unRegisterDispose();
        }
    };
}

File: PaMVPKotlinDemo/lib-core/src/main/java/com/pa/core/base/BaseActivity.kt
package com.pa.core.base


import android.content.DialogInterface.OnCancelListener
import android.os.Bundle
import android.support.v7.app.AppCompatActivity
import com.pa.core.view.CustomProgressDialog


/**
 * Created by wen on 2018/5/14.
 */
abstract class BaseActivity<T : IPresenter> : AppCompatActivity(), IView {

    protected var mPresenter: T? = null

    /**
     * 监听dialog的返回  取消后注销请求
     */
    internal var onCancelListener: OnCancelListener = OnCancelListener { mPresenter?.unRegisterDispose() }

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(getlayoutId())
        //如果转换是不可能的，转换操作符会抛出一个异常。因此，我们称之为不安全的。 Kotlin 中的不安全转换由中缀操作符 as
        //为了避免抛出异常，可以使用安全转换操作符 as?，它可以在失败时返回 null：
        mPresenter = createPresenter() as? T
        initData()
        initView()
        initTitle()
    }

    protected abstract fun getlayoutId(): Int

    protected abstract fun createPresenter(): IPresenter?

    open fun initData() {}

    open fun initView() {}

    open fun initTitle() {}


    override fun showLoadingView() {
        if (!this.isFinishing && !CustomProgressDialog.isShow) {
            CustomProgressDialog.show(this, "登录中...", false, onCancelListener)
        }
    }

    override fun hideLoadingView() {
        if (!this.isFinishing && CustomProgressDialog.isShow) {
            CustomProgressDialog.dimiss()
        }
    }

    override fun onDestroy() {
        if (null != mPresenter) {
            mPresenter?.onDestroy()
        }
        super.onDestroy()
    }
}
